File: backend/src/modules/auth/auth.service.ts
import { Injectable, UnauthorizedException } from '@nestjs/common';
import { JwtService } from '@nestjs/jwt';
import { ConfigService } from '@nestjs/config';
import { PrismaService } from '@common/config/database/prisma.service';
import bcrypt from 'bcrypt';

export interface AdminCreateResult {
  id: string;
  nombre: string;
  email: string;
}

export interface AuthResponse {
  accessToken: string;
  admin: AdminCreateResult;
}

export interface AdminBasic {
  id: string;
  email: string;
  claveHash: string;
  nombre: string;
}

const SALT_ROUNDS = 12;
const MIN_PASSWORD_LENGTH = 8;

@Injectable()
export class AuthService {
  constructor(
    private readonly prisma: PrismaService,
    private readonly jwtService: JwtService,
    private readonly configService: ConfigService,
  ) {}

  private async hashearPassword(password: string): Promise<string> {
    return bcrypt.hash(password, SALT_ROUNDS);
  }

  async validarAdmin(email: string, password: string): Promise<AuthResponse> {
    if (!email || !password) {
      throw new UnauthorizedException('Complete todos los campos');
    }

    const admin = await this.prisma.administrador.findUnique({ where: { email } });
    if (!admin) {
      throw new UnauthorizedException('Email no registrado');
    }

    const passwordMatch = await bcrypt.compare(password, admin.claveHash);
    if (!passwordMatch) {
      throw new UnauthorizedException('Clave incorrecta');
    }

    await this.prisma.administrador.update({
      where: { id: admin.id },
      data: { ultimoAccesoEn: new Date() },
    });

    const payload = { sub: admin.id, email: admin.email, rol: 'admin' as const };

    return {
      accessToken: this.jwtService.sign(payload),
      admin: { id: admin.id, nombre: admin.nombre, email: admin.email },
    };
  }

  async registrarAdmin(nombre: string, email: string, password: string): Promise<AdminCreateResult> {
    const existingAdmin = await this.prisma.administrador.findUnique({ where: { email } });
    if (existingAdmin) {
      throw new Error('El email ya está registrado');
    }

    if (!password || password.length < MIN_PASSWORD_LENGTH) {
      throw new Error('La contraseña debe tener al menos 8 caracteres');
    }

    const hashedPassword = await this.hashearPassword(password);

    return this.prisma.administrador.create({
      data: { nombre, email, claveHash: hashedPassword },
    });
  }

  async cambiarPassword(
    adminId: string,
    currentPassword: string,
    newPassword: string,
  ): Promise<{ mensaje: string }> {
    if (!newPassword || newPassword.length < MIN_PASSWORD_LENGTH) {
      throw new UnauthorizedException('La nueva contraseña debe tener al menos 8 caracteres');
    }

    const admin = await this.prisma.administrador.findUnique({ where: { id: adminId } });
    if (!admin) {
      throw new UnauthorizedException('Administrador no encontrado');
    }

    const passwordMatch = await bcrypt.compare(currentPassword, admin.claveHash);
    if (!passwordMatch) {
      throw new UnauthorizedException('La contraseña actual es incorrecta');
    }

    const newHash = await this.hashearPassword(newPassword);
    await this.prisma.administrador.update({
      where: { id: adminId },
      data: { claveHash: newHash },
    });

    return { mensaje: 'Contraseña actualizada exitosamente' };
  }

  async resetPassword(
    adminId: string,
    nuevaPassword: string,
  ): Promise<{ mensaje: string }> {
    if (!nuevaPassword || nuevaPassword.length < MIN_PASSWORD_LENGTH) {
      throw new UnauthorizedException('La nueva contraseña debe tener al menos 8 caracteres');
    }

    const admin = await this.prisma.administrador.findUnique({ where: { id: adminId } });
    if (!admin) {
      throw new UnauthorizedException('Administrador no encontrado');
    }

    const newHash = await this.hashearPassword(nuevaPassword);
    await this.prisma.administrador.update({
      where: { id: adminId },
      data: { claveHash: newHash },
    });

    return { mensaje: 'Contrasena reseteada exitosamente' };
  }
}

File: backend/src/main/java/com/sistemapedidos/modules/auth/AuthService.java
package com.sistemapedidos.modules.auth;

import com.sistemapedidos.common.error.BadRequestException;
import com.sistemapedidos.common.error.UnauthorizedException;
import com.sistemapedidos.domain.entity.Administrador;
import com.sistemapedidos.domain.repository.AdministradorRepository;
import com.sistemapedidos.modules.auth.dto.AuthResponse;
import com.sistemapedidos.security.JwtService;
import lombok.RequiredArgsConstructor;
import org.springframework.security.crypto.password.PasswordEncoder;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.time.LocalDateTime;
import java.time.ZoneOffset;
import java.util.Map;

@Service
@RequiredArgsConstructor
public class AuthService {

    private static final int MIN_PASSWORD_LENGTH = 8;

    private final AdministradorRepository administradorRepository;
    private final PasswordEncoder passwordEncoder;
    private final JwtService jwtService;

    @Transactional
    public AuthResponse validarAdmin(String email, String password) {
        if (email == null || email.isBlank() || password == null || password.isBlank()) {
            throw new UnauthorizedException("Complete todos los campos");
        }

        Administrador admin = administradorRepository.findByEmail(email)
                .orElseThrow(() -> new UnauthorizedException("Email no registrado"));

        if (!passwordEncoder.matches(password, admin.getClaveHash())) {
            throw new UnauthorizedException("Clave incorrecta");
        }

        admin.setUltimoAccesoEn(LocalDateTime.now(ZoneOffset.UTC));
        administradorRepository.save(admin);

        String token = jwtService.generarToken(admin.getId(), admin.getEmail(), "admin");
        return new AuthResponse(token, new AuthResponse.AdminBasic(admin.getId(), admin.getNombre(), admin.getEmail()));
    }

    @Transactional
    public AuthResponse.AdminBasic registrarAdmin(String nombre, String email, String password) {
        if (administradorRepository.findByEmail(email).isPresent()) {
            throw new BadRequestException("El email ya esta registrado");
        }
        if (password == null || password.length() < MIN_PASSWORD_LENGTH) {
            throw new BadRequestException("La contrasena debe tener al menos 8 caracteres");
        }

        Administrador admin = new Administrador();
        admin.setNombre(nombre);
        admin.setEmail(email);
        admin.setClaveHash(passwordEncoder.encode(password));
        administradorRepository.save(admin);

        return new AuthResponse.AdminBasic(admin.getId(), admin.getNombre(), admin.getEmail());
    }

    @Transactional
    public Map<String, String> cambiarPassword(String adminId, String currentPassword, String newPassword) {
        if (newPassword == null || newPassword.length() < MIN_PASSWORD_LENGTH) {
            throw new UnauthorizedException("La nueva contrasena debe tener al menos 8 caracteres");
        }

        Administrador admin = administradorRepository.findById(adminId)
                .orElseThrow(() -> new UnauthorizedException("Administrador no encontrado"));

        if (!passwordEncoder.matches(currentPassword, admin.getClaveHash())) {
            throw new UnauthorizedException("La contrasena actual es incorrecta");
        }

        admin.setClaveHash(passwordEncoder.encode(newPassword));
        administradorRepository.save(admin);

        return Map.of("mensaje", "Contrasena actualizada exitosamente");
    }

    @Transactional
    public Map<String, String> resetPassword(String adminId, String nuevaPassword) {
        if (nuevaPassword == null || nuevaPassword.length() < MIN_PASSWORD_LENGTH) {
            throw new UnauthorizedException("La nueva contrasena debe tener al menos 8 caracteres");
        }

        Administrador admin = administradorRepository.findById(adminId)
                .orElseThrow(() -> new UnauthorizedException("Administrador no encontrado"));

        admin.setClaveHash(passwordEncoder.encode(nuevaPassword));
        administradorRepository.save(admin);

        return Map.of("mensaje", "Contrasena reseteada exitosamente");
    }
}
